From: Eryu Guan <eguan@redhat.com>
Subject: [PATCH] ext4: fix fdatasync(2) after fallocate(2) operation
Date: Wed, 15 Nov 2017 19:54:19 +0800
Cc: Theodore Ts'o <tytso@mit.edu>, Eryu Guan <eguan@redhat.com>
To: linux-ext4@vger.kernel.org

Currently, fallocate(2) with KEEP_SIZE followed by a fdatasync(2)
then crash, we'll see wrong allocated block number (stat -c %b), the
blocks allocated beyond EOF are all lost. fstests generic/468
exposes this bug.

Commit 67a7d5f561f4 ("ext4: fix fdatasync(2) after extent
manipulation operations") fixed all the other extent manipulation
operation paths such as hole punch, zero range, collapse range etc.,
but forgot the fallocate case.

So similarly, fix it by recording the correct journal tid in ext4
inode in fallocate(2) path, so that ext4_sync_file() will wait for
the right tid to be committed on fdatasync(2).

Signed-off-by: Eryu Guan <eguan@redhat.com>
---

Note that fstests case generic/468 has not been pushed to upstream yet,
please find it at
https://patchwork.kernel.org/patch/7284531/

So the final test seq number can be different, though it's unlikely to
change. If changed I'll send an updated patch.

I've tested this patch with the pending generic/468 test and a auto
group run of fstests with default ext4 configuration (data=ordered and
4k block size), test results all look good.

 fs/ext4/extents.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 07bca11749d4..c941251ac0c0 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -4722,6 +4722,7 @@ static int ext4_alloc_file_blocks(struct file *file, ext4_lblk_t offset,
 						    EXT4_INODE_EOFBLOCKS);
 		}
 		ext4_mark_inode_dirty(handle, inode);
+		ext4_update_inode_fsync_trans(handle, inode, 1);
 		ret2 = ext4_journal_stop(handle);
 		if (ret2)
 			break;
